Singer.API/Services/UserService.cs

```diff
@@ -0,0 +1,147 @@
+      {
+         return await _appContext.Users
+            .OfType<T>()
+            .ToListAsync();
+      }
+
+      public async Task<SearchResults<CareUserDTO>> GetUsersAsync<T>(string sortColumn,
+         string sortDirection,
+         string filter,
+         int page = 0,
+         int userPerPage = 15) where T : CareUser
+      {
+         var users = _appContext.CareUsers.AsQueryable();
+
+         var orderByLambda = PropertyHelpers.GetPropertySelector<CareUserDTO>(sortColumn);
+
+         var result = users.ToPagedList(Filter(filter), ProjectToCareUserDTO(),
+            orderByLambda, sortDirection, page,
+            userPerPage);
+         return result;
+      }
+
+      private static Expression<Func<CareUser, CareUserDTO>> ProjectToCareUserDTO()
+      {
+         return x => new CareUserDTO
+         {
+            Id = x.Id,
+            UserId = x.UserId,
+            Name = x.User.Name,
+            AgeGroup = x.AgeGroup,
+            UserName = x.User.UserName,
+            HasTrajectory = x.HasTrajectory,
+            HasVacationDayCare = x.HasVacationDayCare,
+            BirthDay = x.BirthDay,
+            HasNormalDayCare = x.HasNormalDayCare,
+            HasResources = x.HasResources,
+            CaseNumber = x.CaseNumber,
+            IsExtern = x.IsExtern,
+            Email = x.User.Email
+
+         };
+      }
+
+      public async Task<T> GetUserAsync<T>(Guid id) where T : CareUser
+      {
+         var user = await _appContext.Users.FindAsync(id.ToString()) as T;
+         if (user == null)
+         {
+            throw new UserNotFoundException();
+         }
+
+         return user;
+      }
+
+      public async Task<bool> UpdateUserAsync<T>(T user, Guid id) where T : CareUser
+      {
+         User userToUpdate;
+         try
+         {
+            //Check if id exists
+            userToUpdate = _appContext.Users.Single(u => u.Id == id);
+         }
+         catch
+         {
+            throw new BadInputException();
+         }
+
+         //Ensure client is not trying to change the ID
+         if (user.Id != id)
+         {
+            throw new BadInputException();
+         }
+
+         //Convert user DTO to view
+         userToUpdate = user.User;
+         //userToUpdate.Id = id.ToString();
+
+         //And finally update database
+         _appContext.Users.Update(userToUpdate);
+         await _appContext.SaveChangesAsync();
+         return true;
+      }
+
+      public async Task DeleteUserAsync(Guid id)
+      {
+         User dbUser;
+         try
+         {
+            //Check if id exists
+            dbUser = _appContext.Users.Single(u => u.Id == id);
+         }
+         catch
+         {
+            throw new BadInputException();
+         }
+
+         _appContext.Users.Remove(dbUser);
+         await _appContext.SaveChangesAsync();
+      }
+
+      private static Expression<Func<CareUser, bool>> Filter(string filter)
```

Comment:
Is there a specific reason not to use the extension method IQueryableExtensions.Filter?

Reply:
Yes, sadly I couldn't get the Filter working with a type with subtype properties (like the `CareUser` has got the `User`)